sshproxy.go

```diff
@@ -166,73 +167,89 @@ func (s *Server) handleRequests(reqs <-chan *ssh.Request, channel ssh.Channel, c
 				// append client env to the command environment
 				keyLen := req.Payload[3]
 				valLen := req.Payload[keyLen+7]
-				key := string(req.Payload[4 : keyLen+4])
-				val := string(req.Payload[keyLen+8 : keyLen+valLen+8])
+				key := string(req.Payload[4: keyLen+4])
+				val := string(req.Payload[keyLen+8: keyLen+valLen+8])
 				env = append(env, fmt.Sprintf("%s=%s", key, val))
 			}
 			req.Reply(s.passEnv, nil)
 		case "pty-req":
-			// client has requested a PTY
-			wantsPty = true
-			req.Reply(true, nil)
-		case "shell":
-			// client has not supplied a command, reject
-			req.Reply(false, nil)
+			ok := terminal == nil
+			if ok {
+				var err error
+				terminal, err = pty.NewTerminal()
+				ok = err == nil
+			}
+			if ok {
+				termLen := req.Payload[3]
+				term := req.Payload[4: termLen+4]
+				env = append(env, fmt.Sprintf("TERM=%s", term))
+				w := int(binary.BigEndian.Uint32(req.Payload[termLen+4: termLen+8]))
+				h := int(binary.BigEndian.Uint32(req.Payload[termLen+8: termLen+12]))
+				ok = terminal.SetWinSize(w, h) == nil
+			}
+			req.Reply(ok, nil)
+		case "window-change":
+			if terminal != nil {
+				w := int(binary.BigEndian.Uint32(req.Payload[0:4]))
+				h := int(binary.BigEndian.Uint32(req.Payload[4:8]))
+				terminal.SetWinSize(w, h)
+			}
 		case "exec":
 			// setup is done, parse client exec command
 			cmdLen := req.Payload[3]
-			command := string(req.Payload[4 : cmdLen+4])
-			req.Reply(true, nil)
-			s.handleExec(conn, channel, command, env, wantsPty)
-			log.Printf("Closing SSH channel with %s", conn.RemoteAddr())
-			channel.Close()
-			return
+			command := string(req.Payload[4: cmdLen+4])
+			log.Printf("Handling command '%s' from %s", command, conn.RemoteAddr())
+			cmd := exec.Command(s.shell)
+			cmd.Env = append(cmd.Env, env...)
+			cmd.Env = append(cmd.Env, fmt.Sprintf("SSH_USER=%s", conn.User()))
+			cmd.Env = append(cmd.Env, fmt.Sprintf("SSH_ORIGINAL_COMMAND=%s", command))
+
+			if terminal != nil {
+				err := terminal.Start(cmd)
+				if err != nil {
+					panic(err)
+				}
+
+				go io.Copy(terminal, channel)
+				go io.Copy(channel, terminal)
+			} else {
+				stdout, err := cmd.StdoutPipe()
+				if err != nil {
+					panic(err)
+				}
+
+				stderr, err := cmd.StderrPipe()
+				if err != nil {
+					panic(err)
+				}
+
+				stdin, err := cmd.StdinPipe()
+				if err != nil {
+					panic(err)
+				}
+
+				if err = cmd.Start(); err != nil {
+					panic(err)
+				}
+
+				go io.Copy(stdin, channel)
+				go io.Copy(channel, stdout)
+				go io.Copy(channel.Stderr(), stderr)
+			}
+
+			go func() {
+				cmd.Wait()
```

Comment:
@lekkas yes, it will now [kill the shell process when the channel dies](https://github.com/resin-io/sshproxy/blob/6dbb9c5fba43c486916f0746dbd51551a07c736b/sshproxy.go#L242-L263).